### PR TITLE
WIP - Revamping web services

### PR DESCRIPTION
**Pull request by hypeJunction**
DO NOT MERGE - WORK IN PROGRESS

What this does:
- Starts moving web services to OO
- Removes eval() and parameter serialization in favor of calling callables
- Simplifies parameter validation
- Adds API versioning
- Changes how callback function get their arguments - api 2.0 just passes an associative array instead of each parameter being a separate argument
- [ ] deprecate removed function in 1.x
- [ ] review existing texts
- [ ] add PHPUnit(?) tests for new classes

### CODE REVIEW

**Comment by hypeJunction** [comment]
@Elgg/owners Any suggestions on the how to implement the registry?

File: mod/web_services/start.php [hunk 1]
```diff
@@ -1,338 +1,255 @@
...
+/**
+ * Simple api to return a list of all api's installed on the system.
+ *
+ * @return array
+ * @access private
+ */
+function list_all_apis() {
+
+	$methods = ApiRegistry::all();
+	ksort($methods);
```
Comment (hypeJunction) [inline mod/web_services/start.php]: Need to ksort() for each API version

File: mod/web_services/classes/ApiMethod.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	 * For unknown value types, or casting errors, an exception will be thrown
+	 * For required parameters without input values, an exception will be thrown
+	 *
+	 * @return array
+	 * @throws APIException
+	 * @since 2.0
+	 * @access private
+	 */
+	protected function getParams() {
+		global $API_METHODS;
```
Comment (mrclay) [inline mod/web_services/classes/ApiMethod.php]: I thought this was going away?

File: mod/web_services/classes/ApiRegistry.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+/**
+ * API method registry
+ */
+class ApiRegistry {
+
+	const API_VERSION = '2.0';
+	
+	static $methods = array();
```
Comment (mrclay) [inline mod/web_services/classes/ApiRegistry.php]: Make this whole class dynamic and make _elgg_ws_registry() return a singleton instance.

**Comment by hypeJunction** [comment]
Hm, something is wrong with the diffs...
I am done for today. Let me know what your thoughts are on this @Elgg/owners 
If you think this is acceptable, I will work on the tests.

**Comment by hypeJunction** [comment]
There are few more things I want to do:

- create shortcut routing ```/api/2.0/system.api.list``` with optional ```?format=json|xml```
- deprecate ```elgg_ws_expose_function()``` in favour of:

```php
elgg_ws_register_method('GET', 'system.api.list', array(
  'function' => 'list_all_apis',
  'parameters' => array(),
  'require_api_auth' => false,
  'api_version' => '2.0',
));
```
- make method params filterable with a hook

**Comment by ewinslow** [comment]
Can we have folks pass class names instead of function names. That way they
can be lazy loaded and (eventually) dependency injected?

Also: why explicitly support versioning instead of just having folks do it
manually like elgg_ws_register_method("system.api.list.v2",...)?

Last question: Does this mean we're sticking with rpc instead of doing
something more RESTful?

**Comment by hypeJunction** [comment]
> Can we have folks pass class names instead of function names. That way they
> can be lazy loaded and (eventually) dependency injected?
I imagine it should work with ```__invoke()```. Haven't tried.

> Also: why explicitly support versioning instead of just having folks do it
> manually like elgg_ws_register_method("system.api.list.v2",...)?
Look at Method::execute(). We handle parameters differently. Plus, it will smoothen the transition. Unless, we just want to completely break all existing exposed methods.

> Last question: Does this mean we're sticking with rpc instead of doing
> something more RESTful?
I think we can move to rest with the next iteration. Need to clean up what we have first. I think REST will be realistic when we have better route maps in place.

**Comment by hypeJunction** [comment]
Are plugin simpletests running on CI? I would expect Travis to break by now.

**Comment by ewinslow** [comment]
Guess not...

**Comment by mrclay** [comment]
If we're going to switch to REST anyway, why force devs to rewrite code now? Or will that transition be BC, too?

**Comment by ewinslow** [comment]
Also see this alternative approach that seems to be gaining some traction: https://community.elgg.org/discussion/view/2158924/could-this-be-the-future-of-elgg-data-access-and-ui-rendering. High-level summary: implement RESTful principles at the application level and just use HTTP as the transport.

**Comment by ewinslow** [comment]
I'd prefer to do this in a BC way for 2.x rather than submitting it as a breaking change for 2.0 at this point.

**Comment by hypeJunction** [comment]
I can make it mostly BC, by not passing default api version, so existing methods will continue working. If folks want to switch to assoc array, they can pass version number. I guess that also means we do not maintain api registry api version keys - just using registration order to see which one takes precedence

**Comment by hypeJunction** [comment]
@mrclay, I was thinking rest should be implemented in parallel, not as a replacement ( new plugin perhaps). I think we can eventually transition to REST as well in a BC way - just need to figure out how we will map query params to URL segments.

**Comment by hypeJunction** [comment]
One way to make things RESTful is to create JSON/XML views for all of our resource pages and enable token-based PAM globally. Clients can then use web services to generate an auth token (or use their API token) to just request existing page URLs.
We could expose an additional method to request action tokens, so POST calls can be made directly to /action endpoints.

**Comment by hypeJunction** [comment]
I will break this down into multiple PRs.

**Pull request closed without merging**